crates/pico_macros/src/memo_macro.rs

```diff
@@ -1,42 +1,144 @@
+                            Ok(DbArg(ident))
+                        } else {
+                            Err(DarlingError::custom("Expected string literal")
+                                .with_span(&expr_lit.lit))
+                        }
+                    }
+                    Expr::Path(expr_path) => {
+                        if let Some(segment) = expr_path.path.segments.last() {
+                            Ok(DbArg(segment.ident.clone()))
+                        } else {
+                            Err(DarlingError::custom("Empty path for db")
+                                .with_span(&expr_path.path))
+                        }
+                    }
+                    other => {
+                        Err(DarlingError::custom("Unsupported expression for db").with_span(other))
+                    }
+                }
+            }
+            _ => Err(DarlingError::custom("Unsupported meta for db").with_span(item)),
+        }
+    }
+}
+
+#[derive(Debug, FromMeta)]
+#[darling(derive_syn_parse)]
+struct MemoArgs {
+    #[darling(default)]
+    db: Option<DbArg>,
+}
+
+pub(crate) fn memo_macro(attr: TokenStream, item: TokenStream) -> TokenStream {
+    let args_: MemoArgs = match parse(attr) {
+        Ok(v) => v,
+        Err(e) => {
+            return e.to_compile_error().into();
+        }
+    };
 
-pub(crate) fn memo_macro(_args: TokenStream, item: TokenStream) -> TokenStream {
     let ItemFn {
-        sig,
+        mut sig,
         vis,
-        block,
+        mut block,
         attrs,
     } = parse_macro_input!(item as ItemFn);
-
     let fn_hash = hash(&sig);
 
     if sig.inputs.is_empty() {
         return Error::new_spanned(
             &sig,
-            "Memoized function must have at least one argument (&Database)",
+            "Memoized function must have at least one argument (db or &self)",
         )
         .to_compile_error()
         .into();
     }
 
-    let db_arg = match &sig.inputs[0] {
-        FnArg::Typed(PatType { pat, .. }) => pat,
-        _ => unreachable!(),
+    let db_pos = if let Some(DbArg(ref db_ident)) = args_.db {
+        sig.inputs
+            .iter()
+            .position(|arg| match arg {
+                FnArg::Typed(PatType { pat, .. }) => {
+                    if let Pat::Ident(pi) = &**pat {
+                        pi.ident == *db_ident
+                    } else {
+                        false
+                    }
+                }
+                _ => false,
+            })
+            .unwrap_or(0)
+    } else {
+        sig.inputs
+            .iter()
+            .position(|arg| {
+                matches!(arg, FnArg::Receiver(rcv) if rcv.reference.is_some())
+                    || matches!(arg, FnArg::Typed(_))
+            })
+            .unwrap_or(0)
+    };
+
+    let (db_arg, closure_db_arg) = match &sig.inputs[db_pos] {
+        FnArg::Receiver(rcv) => {
+            if rcv.reference.is_none() {
```

Comment:
should we disallow `&mut self` in the macro? It looks like we rewrite it to be `&self` regardless, which makes for confusing errors:

```
    #[memo]
    fn first_letter_2(&mut self, input_id: SourceId<Input>) -> char {
        self.storage.set(Input {
            key: todo!(),
            value: todo!(),
        });
       
```

```
      #[memo]
|     ^^^^^^^ `__db` is a `&` reference, so the data it refers to cannot be borrowed as mutable
```